pyface/base_application.py

```diff
@@ -0,0 +1,395 @@
+        their start method, and should call the superclass start() method
+        before doing any work themselves.
+        """
+        logger.info('---- Application starting ----')
+        return True
+
+    def stop(self):
+        """ Stop the application, cleanly releasing resources if possible. """
+        logger.info('---- Application stopping ----')
+        return True
+
+    def run(self):
+        """ Run the application
+
+        Return
+        ------
+        status : bool
+            Whether or not the application ran normally
+        """
+        started = run = stopped = False
+
+        # Start up the application.
+        self._fire_application_event('starting')
+        started = self.start()
+        if started:
+
+            logger.info('---- Application started ----')
+            self._fire_application_event('started')
+
+            try:
+                run = self._run()
+            except ApplicationExit as exc:
+                if exc.args == ():
+                    logger.info("---- ApplicationExit raised ----")
+                else:
+                    logger.exception("---- ApplicationExit raised ----")
+                run = (exc.args == ())
+            finally:
+                # Try to shut the application down.
+                self._fire_application_event('stopping')
+                stopped = self.stop()
+                if stopped:
+                    self._fire_application_event('stopped')
+                    logger.info('---- Application stopped ----')
+
+        return started and run and stopped
+
+    def exit(self, force=False):
+        """ Exits the application.
+
+        This method handles a request to shut down the application by the user,
+        eg. from a menu.  If force is False, the application can potentially
+        veto the close event, leaving the application in the state that it was
+        before the exit method was called.
+
+        Parameters
+        ----------
+        force : bool, optional (default False)
+            If set, windows will receive no closing events and will be
+            destroyed unconditionally. This can be useful for reliably tearing
+            down regression tests, but should be used with caution.
+
+        Raises
+        ------
+        ApplicationExit
+            Some subclasses may trigger the exit by raising ApplicationExit.
+        """
+        logger.info('---- Application exit started ----')
+        if force or self._can_exit():
+            try:
+                self._prepare_exit()
+            except Exception:
+                logger.exception("Error preparing for application exit")
+            finally:
+                logger.info('---- Application exit ----')
+                self._exit()
+        else:
+            logger.info('---- Application exit vetoed ----')
+
+    # Initialization utilities -----------------------------------------------
+
+    def setup_logging(self):
```

Comment:
On a second look, it feels to me like we are trying to do too many things here. It doesn't feel like this class' responsibility to create all this logging infrastructure. I would leave this empty and let users subclass and implement if needed.

Reply:
Additionally, I believe that this method is never called. So I would make it empty in this base class, but I would call it at the beginning of the `start` method, so that if someone subclasses it, it is invoked automatically.

Reply:
The intent here was to provide a sane default for setting up logging, but allow the user to call or not call (and override, as desired) in their start method (and similarly with the teardown utilities).  The same idea was there for the excepthook and application home code.

A large part of this was to simplify _testing_ since logging, exception hooks and home directories are annoyingly global and having them as side-effects of start-up and teardown is painful.

Perhaps the right thing to do is to pull them out and make them helper functions (or better, context managers?), so something like:
```
app = Application()
with setup_logging(app), install_excepthook(app):
        app.run()
```
or, I guess,
```
app = Application()
with app.setup_logging(), app.install_excepthook():
        app.run()
```